src/UnitOfWork.php

```
@@ -1781,6 +1781,10 @@ public function addToIdentityMap($entity)
      */
     final public static function getIdHashByIdentifier(array $identifier): string
     {
+        if (array_filter($identifier, 'is_array')) {
+            throw new UnexpectedValueException('Unexpected identifier value: Expecting scalar, got array.');
+        }
```

I think, we can handle this case in the mapper function below. And we could throw that exception if we encounter anything that is not a scalar or stringable.

---

Done in new PR targeting `3.1.x` branch https://github.com/doctrine/orm/pull/11285

---

@derrabus Unfortunately we cannot do that because it breaks the find by custom id object feature :
```
$ vendor/bin/phpunit

1) Doctrine\Tests\ORM\Functional\CustomIdObjectTypeTest::testFindByCustomIdObject
UnexpectedValueException: Unexpected identifier value: Expecting scalar, got object.

2) Doctrine\Tests\ORM\Functional\CustomIdObjectTypeTest::testFetchJoinCustomIdObject
UnexpectedValueException: Unexpected identifier value: Expecting scalar, got object.

3) Doctrine\Tests\ORM\Functional\CustomIdObjectTypeTest::testFetchJoinWhereCustomIdObject
UnexpectedValueException: Unexpected identifier value: Expecting scalar, got object.

4) Doctrine\Tests\ORM\Functional\PaginationTest::testCustomIdTypeWithoutOutputWalker
UnexpectedValueException: Unexpected identifier value: Expecting scalar, got object.

# etc...
```

---

But those custom IDs need to be stringable, otherwise the `implode` below would fail. I don't understand the issue here.

---

In this PR I'm just checking that the "_array_" type is invalid.
So the only check we can do inside the mapper function is to check that `$value` is an array
```php
if (is_array($value)) {
    throw new UnexpectedValueException(...);
}
```
This bit of code is not dealing with other types like objects.

Do you want me to move the `is_array($value)` condition in the mapper function then ?

---

> In this PR I'm just checking that the "_array_" type is invalid.

Which is a bit arbitrary. If we add validation here, let's do it properly and catch everything that we don't accept.

> This bit of code is not dealing with other types like objects.

But it should.

---

So you mean we should do something like
```php
if (! is_scalar($value) && ! ($value instanceof Stringable)) ...
```
?

---

@derrabus I [commited in the new PR targeting 3.1.x](https://github.com/doctrine/orm/pull/11285/commits/121b23621ba89a432be917945cc53a6ba493f80a),  I hope that's what we want 🤞

---

Thank you!